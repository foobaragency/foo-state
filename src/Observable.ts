/**
 * Inspired by rxjs BehaviorSubject
 * https://rxjs.dev/api/index/class/BehaviorSubject
 */

import { isBrowser } from "./isBrowser"

export class Observable<TState> {
  private static stateIndex = 0

  private static eventOptions: AddEventListenerOptions = {
    capture: false,
    once: false,
    passive: true,
  }

  private readonly eventName = `${Observable.stateIndex++}-global-state-change`

  constructor(private _value: TState) {}

  get value() {
    return this._value
  }

  next(value: TState) {
    if (value === this._value) {
      return
    }

    this._value = value

    if (isBrowser()) {
      window.dispatchEvent(new Event(this.eventName))
    }
  }

  subscribe(subscriber: (value: TState) => void) {
    const eventListener = () => subscriber(this._value)
    const eventName = this.eventName

    window.addEventListener(eventName, eventListener, Observable.eventOptions)

    function unsubscribe() {
      window.removeEventListener(
        eventName,
        eventListener,
        Observable.eventOptions
      )
    }

    return { unsubscribe }
  }
}
