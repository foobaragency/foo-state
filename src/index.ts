export * from "./types"
export { createGlobalState } from "./createGlobalState"
