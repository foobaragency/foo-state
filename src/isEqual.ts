import stringify from "safe-stable-stringify"

export function isEqual(value1: unknown, value2: unknown): boolean {
  return stringify(value1) === stringify(value2)
}
