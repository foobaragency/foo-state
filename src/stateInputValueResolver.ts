const isFunction = (value: unknown): value is Function =>
  typeof value === "function"

const stateInputValueResolver = <TState, TValue = TState>(
  currentValue: TState,
  nextValue: TValue | ((prevState: TState) => TValue)
): TValue => {
  return isFunction(nextValue) ? nextValue(currentValue) : nextValue
}

export default stateInputValueResolver
