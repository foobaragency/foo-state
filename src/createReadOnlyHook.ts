import { useEffect, useState } from "react"

import { Observable } from "./Observable"

export function createReadOnlyHook<TState, TPartial>(
  state$: Observable<TState>,
  project: (state: TState) => TPartial
) {

  // We want to set the value as-is and avoid React resolving in case it is a function
  /**
   * If the value passed by the user is a function, we are already resolving it previously.
   * Therefore, we want to set the value as is, and thus, we wrap it in a callback so React resolved to our value.
   */
  const generateProjectedValueCallback = (value: TState) => {
    const projectedValue = project(value)
    const projectedValueCallback = () => projectedValue

    return projectedValueCallback
  }

  return () => {
    const initialValueCallback = generateProjectedValueCallback(state$.initializationValue)
    const [state, setState] = useState(initialValueCallback)

    useEffect(() => {
      /**
       * We use the `initializationValue` and only update the state after mounting the component.
       * This helps us keep a consistent and predictable state between server and client rendering
       *  consequently avoiding a possible hydration mismatch.
       */
      const stateValueAfterMountCallback = generateProjectedValueCallback(state$.value)
      setState(stateValueAfterMountCallback)

      const subscription = state$.subscribe(observedState => {
        const nextValueCallback = generateProjectedValueCallback(observedState)
        setState(nextValueCallback)
      })

      return () => subscription.unsubscribe()
    }, [])

    return state
  }
}
