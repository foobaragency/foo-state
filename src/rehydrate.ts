import { StorageItem } from "./StorageItem"
import { isEqual } from "./isEqual"
import { PersistenceOptions } from "./types"
import { Observable } from "./Observable"

export function rehydrate<TState>(
  state$: Observable<TState>,
  options: PersistenceOptions
) {
  const { key, type } = options
  const storageLayer =
    type === "sessionStorage"
      ? globalThis.sessionStorage
      : globalThis.localStorage

  const storage = new StorageItem<TState>(key, storageLayer)

  const storedValue = storage.get()

  if (storedValue !== undefined && !isEqual(storedValue, state$.value)) {
    state$.next(storedValue)
  }

  state$.subscribe(state => {
    storage.set(state)
  })
}
