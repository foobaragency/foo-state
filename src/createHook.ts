import { SetStateAction } from "react"
import { BehaviorSubject } from "rxjs"

import { createReadOnlyHook } from "./createReadOnlyHook"
import { SetStateOptions } from "./types"

export function createHook<TState>(
  state$: BehaviorSubject<TState>,
  setGlobalState: (
    state: SetStateAction<TState>,
    options?: SetStateOptions
  ) => void
) {
  const useReadOnlyState = createReadOnlyHook(state$, state => state)

  return () => {
    const state = useReadOnlyState()

    return [state, setGlobalState] as const
  }
}
