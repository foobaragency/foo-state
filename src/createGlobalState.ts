import { SetStateAction } from "react"
import { Observable } from "./Observable"
import { isBrowser } from "./isBrowser"
import { isEqual } from "./isEqual"
import { rehydrate } from "./rehydrate"
import {
  GlobalState,
  GlobalStateOptions,
  PartialState,
  SetStateOptions,
} from "./types"
import updater from "./updater"
import { createHook } from "./createHook"
import { createReadOnlyHook } from "./createReadOnlyHook"

export const createGlobalState = <S>(
  initialState: S,
  options?: GlobalStateOptions
): GlobalState<S> => {
  const state$ = new Observable<S>(initialState)

  if (options?.persistence?.enabled && isBrowser()) {
    rehydrate(state$, options?.persistence)
  }

  const getGlobalState = () => state$.value

  const setGlobalState = (
    state: SetStateAction<S>,
    options?: SetStateOptions
  ) => {
    if (options?.deepCompare) {
      if (isEqual(state$.value, state)) {
        return
      }
    }

    state$.next(updater(state$.value, state))
  }

  const useGlobalState = createHook(state$, setGlobalState)

  const useReadOnlyState = createReadOnlyHook(state$, state => state)

  const usePartialState = (project: (state: S) => PartialState<S>) =>
    createReadOnlyHook(state$, project)

  return {
    useGlobalState,
    useReadOnlyState,
    usePartialState,
    getGlobalState,
    setGlobalState,
  }
}
