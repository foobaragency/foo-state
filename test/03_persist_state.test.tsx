import React from "react"

import { render, fireEvent, screen } from "@testing-library/react"

import { Person } from "../examples/03_persist_state/src/Person"

import "@testing-library/jest-dom"

describe("Persist state:", () => {
  beforeEach(() => {
    localStorage.clear()

    jest.clearAllMocks()
  })

  test("change the first name", async () => {
    render(<Person />)

    const firstName = screen.getByLabelText("First Name")

    expect(firstName).toHaveValue("John")

    fireEvent.change(firstName, { target: { value: "Joe" } })

    expect(firstName).toHaveValue("Joe")
    expect(localStorage.getItem("x-storage-key")).toMatchSnapshot()
  })
})
