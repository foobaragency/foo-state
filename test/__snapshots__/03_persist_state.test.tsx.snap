// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Persist state: change the first name 1`] = `"{"age":43,"firstName":"Joe","lastName":"Doe"}"`;
